In [9]:
import graphlab
graphlab.canvas.set_target('ipynb')

In [27]:
sales = graphlab.SFrame('home_data.gl/')
train_data,test_data = sales.random_split(.8, seed=0)

In [28]:
zipcodes = sales.groupby(key_columns="zipcode",
                         operations = {'mean': graphlab.aggregate.MEAN('price')})

In [29]:
zipcodes.sort(['mean'], ascending = False)

Columns:
	zipcode	str
	mean	float

Rows: 70

Data:
+---------+---------------+
| zipcode |      mean     |
+---------+---------------+
|  98039  |   2160606.6   |
|  98004  | 1355927.09779 |
|  98040  | 1194230.00355 |
|  98112  | 1095499.36803 |
|  98102  | 901258.238095 |
|  98109  | 879623.623853 |
|  98105  | 862825.231441 |
|  98006  | 859684.763052 |
|  98119  |  849448.01087 |
|  98005  | 810164.880952 |
+---------+---------------+
[70 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### The zipcode with the highest average sale price is 98039, and the average sale price is $2,160,606.60.

In [30]:
subset = sales[(sales["sqft_living"] > 2000) & (sales["sqft_living"] <= 4000)]

In [31]:
print round((len(subset) * 1.0) / len(sales) * 100.0, 2)

42.19


### So, 42.19% of the houses in our data set have > 2000 and <= 4000 square feet of living space.

In [35]:
my_features = [
    'bedrooms',
    'bathrooms',
    'sqft_living',
    'sqft_lot',
    'floors',
    'zipcode'
]
advanced_features = [
    'bedrooms',
    'bathrooms',
    'sqft_living',
    'sqft_lot',
    'floors',
    'zipcode',
    'condition',
    'grade',
    'waterfront',
    'view',
    'sqft_above',
    'sqft_basement',
    'yr_built',
    'yr_renovated',
    'lat',
    'long',
    'sqft_living15',
    'sqft_lot15'
]

In [39]:
my_features_model = graphlab.linear_regression.create(
    train_data,
    target = "price",
    features = my_features,
    validation_set = None
)
advanced_features_model = graphlab.linear_regression.create(
    train_data,
    target = "price",
    features = advanced_features,
    validation_set = None)

Linear regression:

--------------------------------------------------------

Number of examples          : 17384

Number of features          : 6

Number of unpacked features : 6

Number of coefficients    : 115

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+---------------+

| Iteration | Passes   | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+--------------+--------------------+---------------+

| 1         | 2        | 0.031100     | 3763208.270524     | 181908.848367 |

+-----------+----------+--------------+--------------------+---------------+

SUCCESS: Optimal solution found.

Linear regression:

--------------------------------------------------------

Number of examples          : 17384

Number of features          : 18

Number of unpacked features : 18

Number of coefficients    : 127

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+---------------+

| Iteration | Passes   | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+--------------+--------------------+---------------+

| 1         | 2        | 0.079290     | 3469012.450487     | 154580.940732 |

| 2         | 3        | 0.133536     | 3469012.450673     | 154580.940735 |

+-----------+----------+--------------+--------------------+---------------+

SUCCESS: Optimal solution found.

In [40]:
print my_features_model.evaluate(test_data)
print advanced_features_model.evaluate(test_data)

{'max_error': 3486584.5093818563, 'rmse': 179542.4333126908}
{'max_error': 3556849.4138490623, 'rmse': 156831.11680200786}


### The RMSE is on the test_data for my_features is \$179,542.43. On the advanced_features model, it's \$156,831.12. The difference between the two is \$22,711.31